File: fastApiTest/concurrent_FASTapi.py
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
import time
import pandas as pd
import numpy as np

url = "http://127.0.0.1:7000/infer"
concurrent_requests = [5, 10, 15]

# Prepare dummy input in Triton JSON format
def make_payload():
    data = np.random.rand(1, 3, 224, 224).astype(np.float32).tolist()
    return {
        "inputs": [
            {
                "name": "input",  # should match your ONNX model input name
                "shape": [1, 3, 224, 224],
                "datatype": "FP32",
                "data": data
            }
        ]
    }

def hit_model():
    payload = make_payload()
    start = time.time()
    response = requests.post(url, json=payload)
    end = time.time()

    if response.status_code == 200:
        res_json = response.json()
        return {
            "status_code": response.status_code,
            "output_shape": res_json.get("output_shape", None),
            "inference_time_sec": end - start
        }
    else:
        return {
            "status_code": response.status_code,
            "output_shape": None,
            "inference_time_sec": end - start
        }

all_results = []

for n in concurrent_requests:
    print(f"\nHitting {n} concurrent requests:")
    results = []
    with ThreadPoolExecutor(max_workers=n) as executor:
        futures = [executor.submit(hit_model) for _ in range(n)]
        for future in as_completed(futures):
            results.append(future.result())
    
    # Add concurrency info
    for r in results:
        r["concurrent_requests"] = n
    all_results.extend(results)

    # Optional: print results
    for i, r in enumerate(results):
        print(f"Request {i+1}: status={r['status_code']}, "
              f"time={r['inference_time_sec']:.4f}s, "
              f"output_shape={r['output_shape']}")

# Save to CSV
df = pd.DataFrame(all_results)
df.to_csv("inference_results.csv", index=False)
print("\nResults saved to inference_results.csv")

File: fastApiTest/concurrent_FastApi.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strconv"
	"sync"
	"time"
)

var url = "http://127.0.0.1:7000/infer"
var concurrentRequests = []int{125, 130, 145}

type ResponseData struct {
	StatusCode        int
	OutputShape       string
	InferenceTimeSec  float64
	ConcurrentRequest int
}

// Struct to match expected JSON response
type InferResponse struct {
	OutputShape interface{} `json:"output_shape"`
}

func hitModel() ResponseData {
	start := time.Now()
	resp, err := http.Get(url)
	if err != nil {
		fmt.Println("Request error:", err)
		return ResponseData{StatusCode: 0, OutputShape: "error", InferenceTimeSec: 0}
	}
	defer resp.Body.Close()

	var inferResp InferResponse
	_ = json.NewDecoder(resp.Body).Decode(&inferResp)

	duration := time.Since(start).Seconds()
	return ResponseData{
		StatusCode:       resp.StatusCode,
		OutputShape:      fmt.Sprintf("%v", inferResp.OutputShape),
		InferenceTimeSec: duration,
	}
}

func main() {
	var allResults []ResponseData

	for _, n := range concurrentRequests {
		fmt.Printf("\nHitting %d concurrent requests:\n", n)
		results := make([]ResponseData, n)
		var wg sync.WaitGroup
		wg.Add(n)

		for i := 0; i < n; i++ {
			go func(i int) {
				defer wg.Done()
				results[i] = hitModel()
				results[i].ConcurrentRequest = n
			}(i)
		}
		wg.Wait()

		allResults = append(allResults, results...)

		for i, r := range results {
			fmt.Printf("Request %d: time=%.4fs, output_shape=%s\n", i+1, r.InferenceTimeSec, r.OutputShape)
		}
	}

	// Save to CSV
	file, err := os.Create("inference_results.csv")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	// Header
	writer.Write([]string{"status_code", "output_shape", "inference_time_sec", "concurrent_requests"})

	for _, r := range allResults {
		writer.Write([]string{
			strconv.Itoa(r.StatusCode),
			r.OutputShape,
			fmt.Sprintf("%.6f", r.InferenceTimeSec),
			strconv.Itoa(r.ConcurrentRequest),
		})
	}

	fmt.Println("\nResults saved to inference_results.csv")
}
